source/common/common/status.cc

```diff
@@ -0,0 +1,121 @@
+#include "common/common/status.h"
+
+#include "absl/strings/str_cat.h"
+
+namespace Envoy {
+
+namespace {
+
+const char EnvoyPayloadUrl[] = {"Envoy"};
+
+absl::string_view statusCodeToString(StatusCode code) {
+  switch (code) {
+  case StatusCode::Ok:
+    return absl::OkStatus().ToString();
+  case StatusCode::CodecProtocolError:
+    return "CodecProtocolError";
+  case StatusCode::BufferFloodError:
+    return "BufferFloodError";
+  case StatusCode::PrematureResponseError:
+    return "PrematureResponseError";
+  case StatusCode::CodecClientError:
+    return "CodecClientError";
+  }
+}
+
+struct EnvoyStatusPayload {
+  EnvoyStatusPayload(StatusCode status_code) : status_code_(status_code) {}
+  StatusCode status_code_;
+};
+
+struct PrematureResponsePayload : public EnvoyStatusPayload {
+  PrematureResponsePayload(Http::Code http_code)
+      : EnvoyStatusPayload(StatusCode::PrematureResponseError), http_code_(http_code) {}
+  Http::Code http_code_;
+};
+
+template <typename T> void storePayload(absl::Status& status, const T& payload) {
+  status.SetPayload(
+      EnvoyPayloadUrl,
+      absl::Cord(absl::string_view(reinterpret_cast<const char*>(&payload), sizeof(payload))));
+}
+
+template <typename T = EnvoyStatusPayload> const T* getPayload(const absl::Status& status) {
+  auto payload = status.GetPayload(EnvoyPayloadUrl);
+  RELEASE_ASSERT(payload.has_value(), "Must have payload");
```

Comment:
I think this can be a normal ASSERT per our typical error guidelines, same below.

Reply:
done